srcpkgs/mimalloc/patches/03-libatomic.patch

```diff
@@ -0,0 +1,112 @@
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -203,6 +203,9 @@ else()
+   endif()
+ endif()
+ 
++include("cmake/atomic.cmake")
++list(APPEND mi_libraries ${CMAKE_REQUIRED_LIBRARIES})
++
+ # -----------------------------------------------------------------------------
+ # Install and output names
+ # -----------------------------------------------------------------------------
+--- /dev/null
++++ b/cmake/atomic.cmake
+@@ -0,0 +1,97 @@
++# Based on:
++# https://github.com/llvm/llvm-project/blob/d4dcb55c7050fd908af2378fa551078d859d994f/llvm/cmake/modules/DetermineGCCCompatible.cmake
++# https://github.com/llvm/llvm-project/blob/d4dcb55c7050fd908af2378fa551078d859d994f/llvm/cmake/modules/CheckAtomic.cmake
++
++INCLUDE(CheckCXXSourceCompiles)
++INCLUDE(CheckLibraryExists)
++
++# Determine if the compiler has GCC-compatible command-line syntax.
++
++if(CMAKE_COMPILER_IS_GNUCXX)
++  set(COMPILER_IS_GCC_COMPATIBLE ON)
++elseif( MSVC )
++  set(COMPILER_IS_GCC_COMPATIBLE OFF)
++elseif( "${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang" )
++  set(COMPILER_IS_GCC_COMPATIBLE ON)
++elseif( "${CMAKE_CXX_COMPILER_ID}" MATCHES "Intel" )
++  set(COMPILER_IS_GCC_COMPATIBLE ON)
++endif()
++
++# Sometimes linking against libatomic is required for atomic ops, if
++# the platform doesn't support lock-free atomics.
++
++function(check_working_cxx_atomics varname)
++  set(OLD_CMAKE_REQUIRED_FLAGS ${CMAKE_REQUIRED_FLAGS})
++  set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} -std=c++11")
++  CHECK_CXX_SOURCE_COMPILES("
++#include <atomic>
++std::atomic<int> x;
++std::atomic<short> y;
++std::atomic<char> z;
++int main() {
++  ++z;
++  ++y;
++  return ++x;
++}
++" ${varname})
++  set(CMAKE_REQUIRED_FLAGS ${OLD_CMAKE_REQUIRED_FLAGS})
++endfunction(check_working_cxx_atomics)
++
++function(check_working_cxx_atomics64 varname)
++  set(OLD_CMAKE_REQUIRED_FLAGS ${CMAKE_REQUIRED_FLAGS})
++  set(CMAKE_REQUIRED_FLAGS "-std=c++11 ${CMAKE_REQUIRED_FLAGS}")
++  CHECK_CXX_SOURCE_COMPILES("
++#include <atomic>
++#include <cstdint>
++std::atomic<uint64_t> x (0);
++int main() {
++  uint64_t i = x.load(std::memory_order_relaxed);
++  (void)i;
++  return 0;
++}
```

Comment:
Report this upstream, I find it hard to believe it actually needs 64 bit atomics on 32 bit platforms

Reply:
I don't know if it needs 64 bit atomics on 32 bit platforms. I borrowed this code from `llvm`:
```
+# Based on:
+# https://github.com/llvm/llvm-project/blob/d4dcb55c7050fd908af2378fa551078d859d994f/llvm/cmake/modules/DetermineGCCCompatible.cmake
+# https://github.com/llvm/llvm-project/blob/d4dcb55c7050fd908af2378fa551078d859d994f/llvm/cmake/modules/CheckAtomic.cmake
```

If you think this patch can be made smaller, I welcome the suggestions.

Reply:
Why would you add the patch if it doesn't need 64-bit atomics on 32-bit platforms?

Reply:
@ericonr because it fails to compile on `armv6`. Saying something about `atomic_add8` (or something similar). So apparently it wants 64bit atomics. See https://github.com/microsoft/mimalloc/pull/429